migration,jobs: refactor long-running migrations and hook up job

ajwerner opened the pull request:
This PR comes in three commits. The first commit reworks the package structure of long-running migrations a bit. The second introduces a migration job. The third introduces a `system.migrations` table to store completion state for migrations. The idea is that running long-running migrations in a job is handy because it provides leasing, observability, and control to users. 

Fixes #58183.

Commit:
migration: rework migration subsystem into multiple packages

No tests were harmed in this commit. Some abstractions which felt like cruft
were thinned.

Release note: None

Commit:
jobs,migration: introduce long running migration job

This commit introduces a job to run long-running migration. This empowers
long-running migrations with leases, pausability, and cancelability.

Fixes #58183

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/59760)
<!-- Reviewable:end -->

tbg reviewed (commented):
Thanks for getting this out!

> Reviewed 30 of 30 files at r1, 23 of 23 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-:-MSdaxsd1NAqIgG91BNt:bdlh69c)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @irfansharif)

___
*[pkg/migration/migrationmanager/manager_external_test.go, line 102 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MSdaZWP3pYz8ZWOyq9w:-MSdaZWP3pYz8ZWOyq9x:by5cw1e) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d9813eb7e5bee27fd51ea7278a1737301b9c8cb/pkg/migration/migrationmanager/manager_external_test.go#L102)):*
> ```Go
>              WHERE (
>                     crdb_internal.pb_to_json(
>                         'cockroach.sql.jobs.jobspb.Payload',
> ```

wow, this pb builtin stuff has become wonderful
___
*[pkg/migration/migrationmanager/manager.go, line 45 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MSdX21O7-mNa08BDPS8:-MSdX21O7-mNa08BDPS9:bqzh5p) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d9813eb7e5bee27fd51ea7278a1737301b9c8cb/pkg/migration/migrationmanager/manager.go#L45)):*
> ```Go
> // NewManager constructs a new Manager.
> //
> // TODO(irfansharif): We'll need to eventually plumb in on a lease manager here.
> ```

Is this TODO obsolete?
___
*[pkg/migration/migrationmanager/manager.go, line 280 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MSdXyph7uwQZFwuz4Ly:-MSdXyph7uwQZFwuz4Lz:b-pynm9a) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d9813eb7e5bee27fd51ea7278a1737301b9c8cb/pkg/migration/migrationmanager/manager.go#L280)):*
> ```Go
>   WHERE status IN ` + jobs.NonTerminalStatusTupleString + `
> 	)
> 	WHERE pl->'longRunningMigration'->'clusterVersion' = $1::JSON;`
> ```

So I understand that if I invoke two overlapping bumps to `21.1` that they will serialize on each other, but what if I do `20.2-15` and `21.1`? They will step through a prefix of the same versions, but it looks like they'll be allowed to run in parallel. I don't know how much we need to do to prevent this, but it'd be good to document the behavior. It could backfire if we moved to more rapid "major" releases or extended the allowed gap between major releases (which is, right now, at zero). In both cases we could end up with concurrent upgrades if the cluster got upgraded in rapid succession.
For now, maybe good enough to return an error?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59760)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @irfansharif, and @tbg)

___
*[pkg/migration/migrationmanager/manager.go, line 45 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MSdX21O7-mNa08BDPS8:-MSfLuMAA3jMqC_meWzr:bo6apac) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d9813eb7e5bee27fd51ea7278a1737301b9c8cb/pkg/migration/migrationmanager/manager.go#L45)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Is this TODO obsolete?
</blockquote></details>

Indeed.
___
*[pkg/migration/migrationmanager/manager.go, line 280 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MSdXyph7uwQZFwuz4Ly:-MSfM2J-1DoFIxvpjKXt:ba7b77f) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d9813eb7e5bee27fd51ea7278a1737301b9c8cb/pkg/migration/migrationmanager/manager.go#L280)):*
>For now, maybe good enough to return an error?

In what scenario are you suggesting we return an error? Right now it effectively allows the concurrent migrations to occur with the caveat that no individual migration will be occurring concurrently. If these things are idempotent and the cluster version never moves backwards (which it shouldn't), then this isn't so hazardous. Sure, there's a small window when these concurrent operations might happen but once any of them succeeds in setting the new cluster version, once any other concurrent attempts finish, we shouldn't see the migration run again.

You have highlighted a big, important, missing thing here. If we do have clients concurrently trying to jump through a number of versions and it turns out that one of those migrations is buggy or not finishing, then we might be in a bit of a pickle where we re-run migrations potentially a lot of times.

Ideally we'd make the successful completion of any migration store that fact somewhere so that the migration doesn't get run again. Today there's this weird thing that the side-effect of setting the cluster version is handled very far away from this code which actually does the migration and only sets it to the final version after running all of the migrations. 

We've got some options:

1) Detect succeeded jobs here and then not run again.

This approach would probably be okay and is very simple but it isn't perfect. Job records get GC'd and it's rather implicit.

2) Write to a key at the end of the job which we read either here or in the job itself.

I'm fine with that but it's a bit heavy. We'd probably want a table. I don't think we're all too keen on random keys like the existing migration keys which are used in `sqlmigrations` today (right?)

3) Rework the version setting code to always step through the versions and commit them one-by-one.

We've disallowed explicit transactions for setting this cluster setting. It makes it so that you might end up only partially getting the version to where you want, which might be weird. I think this is my preferred approach but I could be convinced of one of the others. 


<!-- Sent from Reviewable.io -->

Commit:
sql,systemschema,migration: add `system.migrations`, adopt

This commit adds a table, `system.migrations`, to store the completion
status of long-running migrations. This table is used to ensure
safety and isolation in the face of concurrent attempts to upgrade the
cluster version.

An important facet of this change is the migration to add this table
to the cluster during upgrades. Given that the implementation of long-running
migrations relies on the table's existence, all versions which are associated
with long-running migrations must follow the version which introduces this
table. This migration needs to utilize the new infrastructure rather than
the `sqlmigrations` infrastructure because of some really bad behavior
introduced in 20.2 (#59850). Given two such versions already existed, they
have been assigned new version keys and their old version keys and values
are now merely placeholders. This provides comptability with the earlier
alphas.

Release note: None

ajwerner reviewed (commented):
We had a fun bootstrapping problem to make this all work but I took care of it. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-:-MSk_6hE1Cwv6dQJhUm2:b-v9z4t0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @irfansharif, and @tbg)

___
*[pkg/migration/migrationmanager/manager.go, line 280 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MSdXyph7uwQZFwuz4Ly:-MSksVgt2Bxy-VSpxUKk:b-k2gn1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d9813eb7e5bee27fd51ea7278a1737301b9c8cb/pkg/migration/migrationmanager/manager.go#L280)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

>For now, maybe good enough to return an error?

In what scenario are you suggesting we return an error? Right now it effectively allows the concurrent migrations to occur with the caveat that no individual migration will be occurring concurrently. If these things are idempotent and the cluster version never moves backwards (which it shouldn't), then this isn't so hazardous. Sure, there's a small window when these concurrent operations might happen but once any of them succeeds in setting the new cluster version, once any other concurrent attempts finish, we shouldn't see the migration run again.

You have highlighted a big, important, missing thing here. If we do have clients concurrently trying to jump through a number of versions and it turns out that one of those migrations is buggy or not finishing, then we might be in a bit of a pickle where we re-run migrations potentially a lot of times.

Ideally we'd make the successful completion of any migration store that fact somewhere so that the migration doesn't get run again. Today there's this weird thing that the side-effect of setting the cluster version is handled very far away from this code which actually does the migration and only sets it to the final version after running all of the migrations. 

We've got some options:

1) Detect succeeded jobs here and then not run again.

This approach would probably be okay and is very simple but it isn't perfect. Job records get GC'd and it's rather implicit.

2) Write to a key at the end of the job which we read either here or in the job itself.

I'm fine with that but it's a bit heavy. We'd probably want a table. I don't think we're all too keen on random keys like the existing migration keys which are used in `sqlmigrations` today (right?)

3) Rework the version setting code to always step through the versions and commit them one-by-one.

We've disallowed explicit transactions for setting this cluster setting. It makes it so that you might end up only partially getting the version to where you want, which might be weird. I think this is my preferred approach but I could be convinced of one of the others. 

</blockquote></details>

I went with 2. It was heavier but I couldn't justify lowering the cluster setting changing logic. I think this worked reasonably well. I added a test. The hiccup was the migration but I think I dealt with it just fine. 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
The bootstrap problems here were actually rather thorny but they have been navigated. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-:-MSn9fEcAJAncU5aEOyo:b-iengpe)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @irfansharif, @pbardea, and @tbg)



<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
@pbardea I wonder if you have thoughts on the backup stuff. Should we backup cluster migration state? What sort of enforcement or invariants do we have about restoring backups and their relationship to cluster versions? I'd hope that we write down a version in backup and only ever restore into a cluster with a newer active version. @dt if you've got feelings, I'm interested. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-:-MSnFgcKA22v_-BF4JA3:b-yvzluf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @irfansharif, @pbardea, and @tbg)



<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
This all looks good to me, thanks for all of this work! Since a lot of the nitty gritty is in the jobs, I would ask you to get a second review for that part from someone more experienced with them. I had a few comments, but nothing dramatic.

> Reviewed 25 of 25 files at r3, 24 of 24 files at r4, 24 of 24 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-:-MT0JWD31-COLb4m84fC:bfceu99)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @irfansharif, and @pbardea)

___
*[pkg/clusterversion/cockroach_versions.go, line 206 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT0EBIqDnfE1ZkaGrxP:-MT0EBIqDnfE1ZkaGrxQ:b-w5f3j1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c31b48bc95edbb88a24a718398b7d745a5d7c76/pkg/clusterversion/cockroach_versions.go#L206)):*
> ```Go
> 	// ReplicaVersions enables the versioning of Replica state.
> 	ReplicaVersions
> 	// ReplacedTruncatedAndRangeAppliedStateMigration stands in for
> ```

Can you unexport these to avoid confusion? They shouldn't be referenced from outside this package, right?
___
*[pkg/migration/migration.go, line 177 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT0FJGu8cgTNmLQ9-YG:-MT0FJGu8cgTNmLQ9-YH:bjsd0cs) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c31b48bc95edbb88a24a718398b7d745a5d7c76/pkg/migration/migration.go#L177)):*
> ```Go
> // SQL layer. Lower-level migrations may depend solely on the Cluster.
> type SQLDeps struct {
> 	Cluster  Cluster
> ```

Cluster should split into two subinterfaces, right?

```go
type SystemTenantCluster interface {
  AnyTenantCluster
  ForEveryNode(_ context.Context, op string, func(context.Context, serverpb.MigrationClient)
  UntilClusterStable(/* ... */)
}

type AnyTenantCluster interface {
  DB() *kv.DB
  IterateRangeDescriptors(/* ... */)
}
```

That way SQLMigrationFn can be forced to take only an AnyTenantCluster and thus "ensure" (to a reasonable degree) that folks won't accidentally write migrations that don't work in multi-tenancy. (Obviously we need a mixed version test anyway but still).

I realize you're not working on that issue now. Just throwing it out there. You can punt, I'm already so graceful you're doing all of this jobs work.
___
*[pkg/migration/migration.go, line 182 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT0F03i4GyRm_97tQRe:-MT0F03i4GyRm_97tQRf:b6eam1h) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c31b48bc95edbb88a24a718398b7d745a5d7c76/pkg/migration/migration.go#L182)):*
> ```Go
> }
> 
> type KVMigrationFn func(context.Context, clusterversion.ClusterVersion, Cluster) error
> ```

Don't these need comments for the linter?
___
*[pkg/migration/migration.go, line 183 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT0En4e3r8koxFb_1Kt:-MT0En4e3r8koxFb_1Ku:b2fbqp3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c31b48bc95edbb88a24a718398b7d745a5d7c76/pkg/migration/migration.go#L183)):*
> ```Go
> 
> type KVMigrationFn func(context.Context, clusterversion.ClusterVersion, Cluster) error
> type SQLMigrationFn func(context.Context, clusterversion.ClusterVersion, SQLDeps) error
> ```

:heart:
Commenting to tag the issue: https://github.com/cockroachdb/cockroach/issues/48436#issuecomment-749552817
___
*[pkg/migration/migrationjob/migration_job.go, line 42 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT0GdqW7Wk6F_wp0_Ks:-MT0GdqW7Wk6F_wp0_Kt:b-gdbhdu) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c31b48bc95edbb88a24a718398b7d745a5d7c76/pkg/migration/migrationjob/migration_job.go#L42)):*
> ```Go
> func NewRecord(version clusterversion.ClusterVersion, user security.SQLUsername) jobs.Record {
> 	return jobs.Record{
> 		Description: "Long running migration",
> ```

Is it supposed to be "Long-running"?
___
*[pkg/migration/migrationjob/migration_job.go, line 58 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT0Giyi3GEp44G2hvY-:-MT0Giyi3GEp44G2hvY0:bg8m663) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c31b48bc95edbb88a24a718398b7d745a5d7c76/pkg/migration/migrationjob/migration_job.go#L58)):*
> ```Go
> var _ jobs.Resumer = (*resumer)(nil)
> 
> func (r resumer) Resume(ctx context.Context, execCtxI interface{}) error {
> ```

Since I'm not an expert on jobs, you probably want to get a second pair of eyes on the jobs-related code.
___
*[pkg/migration/migrationjob/migration_job.go, line 67 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT0Gt1l446P9J74stTg:-MT0Gt1l446P9J74stTh:b9zyrxr) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c31b48bc95edbb88a24a718398b7d745a5d7c76/pkg/migration/migrationjob/migration_job.go#L67)):*
> ```Go
> 	alreadyCompleted, err := CheckIfMigrationCompleted(ctx, nil /* txn */, ie, cv)
> 	if alreadyCompleted || err != nil {
> 		return errors.Wrapf(err, "checking migration completion for %v", cv)
> ```

I'm always on the fence about this pattern (using `errors.Wrap` to return nil) but I don't know exactly why, so just a remark and not a suggestion to change.
___
*[pkg/migration/migrationjob/migration_job.go, line 76 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT0H5Hq92Dx0_0XEYdw:-MT0H5Hq92Dx0_0XEYdx:bydu7aw) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c31b48bc95edbb88a24a718398b7d745a5d7c76/pkg/migration/migrationjob/migration_job.go#L76)):*
> ```Go
> 	}
> 	if !ok {
> 		// TODO(ajwerner): Consider treating this as an assertion failure. Jobs
> ```

:thumbsup: could there be exceptions though if we reorder versions?
___
*[pkg/migration/migrationmanager/manager_external_test.go, line 248 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT0IJ43Asy1yQ50QgUN:-MT0IJ43Asy1yQ50QgUO:b8dp3bf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c31b48bc95edbb88a24a718398b7d745a5d7c76/pkg/migration/migrationmanager/manager_external_test.go#L248)):*
> ```Go
> 	defer log.Scope(t).Close(t)
> 
> 	// We're going to be migrating from startKey to endKey. We end up needing
> ```

This is going to be annoying every time we phase out old versions. Could you add a testing knob that can mock out the call to `ListBetween` in `(*Manager).Migrate`? There's already the corresponding thing (though unfortunately as a singleton, but whatever) in TestingRegisterMigrationInterceptor which you are already using.
___
*[pkg/migration/migrationmanager/manager_external_test.go, line 275 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT0InkP7TaMOj4ntW7k:-MT0InkP7TaMOj4ntW7l:b-650132) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c31b48bc95edbb88a24a718398b7d745a5d7c76/pkg/migration/migrationmanager/manager_external_test.go#L275)):*
> ```Go
> 					ctx context.Context, cv clusterversion.ClusterVersion, c migration.SQLDeps,
> 				) error {
> 					migrationRunCounts[cv]++
> ```

You could also `int active int32 // atomic` above, and here

```go
if atomic.AddInt32(&active, 1) != 1 {
  t.Error("unexpected concurrency")
}
time.Sleep(5*time.Millisecond)
atomic.AddInt32(&active, -1)
```

Just anticipating that this test won't be run under the race detector forever :S (Plus, not sure how efficient the race detector is for this particular race)
___
*[pkg/migration/migrationmanager/manager.go, line 294 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT0Hj1R-XuCyxS2rx05:-MT0Hj1R-XuCyxS2rx06:b-fsw10j) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c31b48bc95edbb88a24a718398b7d745a5d7c76/pkg/migration/migrationmanager/manager.go#L294)):*
> ```Go
>   WHERE status IN ` + jobs.NonTerminalStatusTupleString + `
> 	)
> 	WHERE pl->'longRunningMigration'->'clusterVersion' = $1::JSON;`
> ```

Is there a danger of us renaming those fields and badly breaking things? Do these fields have explicit json name tags that don't depend on proto name changes? Relying on proto field names always scares me because we always assume these can be mutated.
I assume there's no way for this protoreflect stuff to use field numbers as accessors?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
With this PR and the associate issues, it feels like this subsystem can and should totally subsume `pkg/sqlmigrations`. I can't think of a reason why `sqlmigrations` should continue to exist. 

@miretskiy any chance I coax you into looking at the job bits of this? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-:-MT1642O7_28TnFCN6Lq:byx5s21)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @irfansharif, and @tbg)

___
*[pkg/clusterversion/cockroach_versions.go, line 206 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT0EBIqDnfE1ZkaGrxP:-MT1Ct7v2EHddSDvyq3E:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c31b48bc95edbb88a24a718398b7d745a5d7c76/pkg/clusterversion/cockroach_versions.go#L206)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Can you unexport these to avoid confusion? They shouldn't be referenced from outside this package, right?
</blockquote></details>

Done.
___
*[pkg/migration/migration.go, line 177 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT0FJGu8cgTNmLQ9-YG:-MT15ggi85ToYuWw5_VW:bca7mtc) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c31b48bc95edbb88a24a718398b7d745a5d7c76/pkg/migration/migration.go#L177)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Cluster should split into two subinterfaces, right?

```go
type SystemTenantCluster interface {
  AnyTenantCluster
  ForEveryNode(_ context.Context, op string, func(context.Context, serverpb.MigrationClient)
  UntilClusterStable(/* ... */)
}

type AnyTenantCluster interface {
  DB() *kv.DB
  IterateRangeDescriptors(/* ... */)
}
```

That way SQLMigrationFn can be forced to take only an AnyTenantCluster and thus "ensure" (to a reasonable degree) that folks won't accidentally write migrations that don't work in multi-tenancy. (Obviously we need a mixed version test anyway but still).

I realize you're not working on that issue now. Just throwing it out there. You can punt, I'm already so graceful you're doing all of this jobs work.
</blockquote></details>

Good point. I don't even think that `IterateRangeDescriptors` is needed for the sql migrations. I've reworked this a bit. It's not amazing but I think it's better. 

I think that getting the tenant work finished here isn't too bad. One of the biggest blockers seems to be https://github.com/cockroachdb/cockroach/pull/58362 which I'm planning on sprucing up this week. 
___
*[pkg/migration/migration.go, line 182 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT0F03i4GyRm_97tQRe:-MT19WXC3sSS60_RDCor:b60r506) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c31b48bc95edbb88a24a718398b7d745a5d7c76/pkg/migration/migration.go#L182)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Don't these need comments for the linter?
</blockquote></details>

Yes, done.
___
*[pkg/migration/migrationjob/migration_job.go, line 42 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT0GdqW7Wk6F_wp0_Ks:-MT19aIR9yfEzVAQjK8C:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c31b48bc95edbb88a24a718398b7d745a5d7c76/pkg/migration/migrationjob/migration_job.go#L42)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Is it supposed to be "Long-running"?
</blockquote></details>

Done.
___
*[pkg/migration/migrationjob/migration_job.go, line 67 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT0Gt1l446P9J74stTg:-MT19fk39vU-YBDu-RGW:b-ey06vt) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c31b48bc95edbb88a24a718398b7d745a5d7c76/pkg/migration/migrationjob/migration_job.go#L67)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I'm always on the fence about this pattern (using `errors.Wrap` to return nil) but I don't know exactly why, so just a remark and not a suggestion to change.
</blockquote></details>

I've come around to using it. I find myself really happy to have annotations indicating the call stack of an error and for various reasons we don't tend to get stack traces back to clients. It should be plenty cheap for everything but very hot code paths. 
___
*[pkg/migration/migrationjob/migration_job.go, line 76 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT0H5Hq92Dx0_0XEYdw:-MT19wdq6_7syAnVCp_A:biol9g) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c31b48bc95edbb88a24a718398b7d745a5d7c76/pkg/migration/migrationjob/migration_job.go#L76)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

:thumbsup: could there be exceptions though if we reorder versions?
</blockquote></details>

I suppose, yes. Added more commentary. 
___
*[pkg/migration/migrationmanager/manager_external_test.go, line 248 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT0IJ43Asy1yQ50QgUN:-MT4Y-zeBSA7n0xyQcgG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c31b48bc95edbb88a24a718398b7d745a5d7c76/pkg/migration/migrationmanager/manager_external_test.go#L248)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This is going to be annoying every time we phase out old versions. Could you add a testing knob that can mock out the call to `ListBetween` in `(*Manager).Migrate`? There's already the corresponding thing (though unfortunately as a singleton, but whatever) in TestingRegisterMigrationInterceptor which you are already using.
</blockquote></details>

Done.
___
*[pkg/migration/migrationmanager/manager_external_test.go, line 275 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT0InkP7TaMOj4ntW7k:-MT4Y0QA0zrt_pAavtcV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c31b48bc95edbb88a24a718398b7d745a5d7c76/pkg/migration/migrationmanager/manager_external_test.go#L275)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

You could also `int active int32 // atomic` above, and here

```go
if atomic.AddInt32(&active, 1) != 1 {
  t.Error("unexpected concurrency")
}
time.Sleep(5*time.Millisecond)
atomic.AddInt32(&active, -1)
```

Just anticipating that this test won't be run under the race detector forever :S (Plus, not sure how efficient the race detector is for this particular race)
</blockquote></details>

Done.
___
*[pkg/migration/migrationmanager/manager.go, line 294 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT0Hj1R-XuCyxS2rx05:-MT1AZwdAhge7u7b_sc1:b1p5u7d) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c31b48bc95edbb88a24a718398b7d745a5d7c76/pkg/migration/migrationmanager/manager.go#L294)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Is there a danger of us renaming those fields and badly breaking things? Do these fields have explicit json name tags that don't depend on proto name changes? Relying on proto field names always scares me because we always assume these can be mutated.
I assume there's no way for this protoreflect stuff to use field numbers as accessors?
</blockquote></details>

Certainly there is. This code needs to stay in sync with the field naming compiled into it. Unit testing here should catch issues. It is somewhat brittle.Fortunately this isn't a mixed version problem. We made these `pb_to_json` functions not eligible for distsql for exactly this reason. 


<!-- Sent from Reviewable.io -->

miretskiy reviewed (approved):
> Reviewed 1 of 30 files at r1, 1 of 25 files at r3, 4 of 44 files at r6, 7 of 22 files at r7, 21 of 47 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59760)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @irfansharif, and @tbg)

___
*[pkg/ccl/backupccl/system_schema.go, line 232 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT6dbzy4h7S7L361UNy:-MT6dbzy4h7S7L361UNz:b-bcxc78) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdd04f8e5bcfc50b46794910fea6dbcc1fa841c1/pkg/ccl/backupccl/system_schema.go#L232)):*
> ```Go
> 	systemschema.LongRunningMigrationsTable.GetName(): {
> 		includeInClusterBackup: optOutOfClusterBackup,
> 	},
> ```

fyi: you'll need to rebase.  This code moved to backupbase package.
___
*[pkg/migration/migration.go, line 17 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT6VHVV-aQtHsmwYB7x:-MT6VHVV-aQtHsmwYB7y:bwny6q2) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdd04f8e5bcfc50b46794910fea6dbcc1fa841c1/pkg/migration/migration.go#L17)):*
> ```
> activate 
> ```
made active?
___
*[pkg/migration/migration.go, line 21 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT6VPGH4qHjKNK3sEP2:-MT6VPGH4qHjKNK3sEP3:b-py24s0) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdd04f8e5bcfc50b46794910fea6dbcc1fa841c1/pkg/migration/migration.go#L21)):*
> ```
> pkg/sqlmigrations.
> ```
don't you want to get rid of it eventually? Perhaps drop mention of this?
___
*[pkg/migration/migrationcluster/cluster.go, line 74 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT6mQ9q9klOtObUbetF:-MT6mQ9r66tOLjbEtZLL:b-xuui2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdd04f8e5bcfc50b46794910fea6dbcc1fa841c1/pkg/migration/migrationcluster/cluster.go#L74)):*
> ```Go
> 	}
> 
> 	// We'll want to rate limit outgoing RPCs (limit pulled out of thin air).
> ```

i'm curious why do we want to rate limit here?  
___
*[pkg/migration/migrationjob/migration_job.go, line 78 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT6ZTBO8R54JyrG8xoG:-MT6ZTBO8R54JyrG8xoH:b-gql3li) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdd04f8e5bcfc50b46794910fea6dbcc1fa841c1/pkg/migration/migrationjob/migration_job.go#L78)):*
> ```Go
> 		// with a newer version where the migration has been re-ordered to be later.
> 		// This should only happen between alphas but is theoretically not illegal.
> 		return nil
> ```

So, that would mark the job completed?
___
*[pkg/migration/migrationjob/migration_job.go, line 99 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT6XF-H62MboZlvQ-mS:-MT6XF-H62MboZlvQ-mT:b-tofqxg) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdd04f8e5bcfc50b46794910fea6dbcc1fa841c1/pkg/migration/migrationjob/migration_job.go#L99)):*
> ```Go
> 	// no-op and new jobs are not created.
> 	return errors.Wrapf(markMigrationCompleted(ctx, ie, cv),
> 		"marking migration complete for %v", cv)
> ```

nit: perhaps long hand?
I see you already discussed this... I don't know; most of the code does "if err != nil { return wrapf }"
___
*[pkg/migration/migrationmanager/manager.go, line 269 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT6e6Zd2a4Nyo-_dkb0:-MT6e6Zd2a4Nyo-_dkb1:bseaci7) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdd04f8e5bcfc50b46794910fea6dbcc1fa841c1/pkg/migration/migrationmanager/manager.go#L269)):*
> ```
> migrationjob.CheckIfMigrationCompleted
> ```

nit: this almost feels like this method might have a better home inside migration package (or migrationmanager) instead of migrationjob?
___
*[pkg/migration/migrationmanager/manager.go, line 302 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT6eNXGEhXKtcKiM5Rd:-MT6eNXGEhXKtcKiM5Re:bjwa7hh) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdd04f8e5bcfc50b46794910fea6dbcc1fa841c1/pkg/migration/migrationmanager/manager.go#L302)):*
> ```Go
> 		SELECT id,
> 		status,
> 		crdb_internal.pb_to_json(
> ```

:tada: 
___
*[pkg/migration/migrationmanager/manager.go, line 310 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT6iV8B9wvtrzjjc66F:-MT6iV8B9wvtrzjjc66G:bd5jsfi) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdd04f8e5bcfc50b46794910fea6dbcc1fa841c1/pkg/migration/migrationmanager/manager.go#L310)):*
> ```Go
>   WHERE status IN ` + jobs.NonTerminalStatusTupleString + `
> 	)
> 	WHERE pl->'longRunningMigration'->'clusterVersion' = $1::JSON;`
> ```

do you need to use cte here? Can't we just 
select id, status, crdb_internal.pb_to_json() pl
where status in ... AND pl->....?
___
*[pkg/migration/migrationmanager/manager.go, line 341 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT6i49U3rI3di7kRObd:-MT6i49U3rI3di7kRObe:b-iws0gb) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdd04f8e5bcfc50b46794910fea6dbcc1fa841c1/pkg/migration/migrationmanager/manager.go#L341)):*
> ```Go
> 		}
> 		log.Errorf(ctx, "%s", buf)
> 		return false, 0, errors.AssertionFailedf("%s", buf)
> ```

will the cluster ever recover from this? 
___
*[pkg/migration/migrationmanager/manager.go, line 360 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT6jDU76KEqcBhu907A:-MT6jDU76KEqcBhu907B:brdtvih) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdd04f8e5bcfc50b46794910fea6dbcc1fa841c1/pkg/migration/migrationmanager/manager.go#L360)):*
> ```Go
> 		// do something similar would be caught at the sql layer (also tested in
> 		// the same logictest). We'll just explicitly append the target version
> 		// here instead, so that we're able to actually migrate into it.
> ```

just wondering: can this be made more explicit? Like, perhaps tests (or test cluster?) can set a variable here that this manager is running under test and only then allow this? Otherwise panic or something?

Might be more trouble than worth though. 


<!-- Sent from Reviewable.io -->

irfansharif commented on pkg/migration/migrationmanager/manager.go:
```diff
@@ -234,3 +234,131 @@ func (m *Manager) Migrate(ctx context.Context, from, to clusterversion.ClusterVe
+		// the same logictest). We'll just explicitly append the target version
+		// here instead, so that we're able to actually migrate into it.
+		versions = []clusterversion.ClusterVersion{to}
+	}
```
Is there a need for this special case now that we have a way to override what's returned through the testing knob?

ajwerner replied:
This is gone now.

irfansharif commented on pkg/migration/migrationmanager/manager.go:
```diff
@@ -78,38 +102,14 @@ func (m *Manager) Migrate(ctx context.Context, from, to clusterversion.ClusterVe
-			if err := migration.Run(ctx, h); err != nil {
-				return err
-			}
+		log.Infof(ctx, "stepping through %s to %s (%s)", from, to, clusterVersions)
```
Do we want to log the specific `clusterVersion` instead of the list of _all_ cluster versions?

ajwerner replied:
I think I just fixed this.

irfansharif commented on pkg/migration/migration.go:
```diff
@@ -0,0 +1,259 @@
+	//   (e) We'll continue to loop around until the node ID list stabilizes
+	//
+	// [*]: We can be a bit more precise here. What UntilClusterStable gives us is a
+	// strict causal happenedbefore relation between running the given closure and
```
"happened-before"?

ajwerner replied:
Done.

irfansharif commented on pkg/migration/migration.go:
```diff
@@ -0,0 +1,259 @@
+// [4]: pkg/kv/kvserver/batch_eval/cmd_migrate.go
+//
+type Migration interface {
+	ClusterVersion() clusterversion.ClusterVersion
```
Given that we want only two real types to implement this, KVMigration and SQLMigration, should we define a hidden method here to restrict the list of types implicitly implementing the Migration interface?

ajwerner replied:
Done.

irfansharif commented on pkg/migration/migrations/long_running_migrations.go:
```diff
@@ -0,0 +1,32 @@
+	"github.com/cockroachdb/cockroach/pkg/sqlmigrations"
+)
+
+func init() {
```
Remove the empty `init`?

ajwerner replied:
Done.

ajwerner reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-:-MT7ehCu3uFvNeQyCHSX:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @irfansharif, @miretskiy, and @tbg)

___
*[pkg/migration/migration.go, line 17 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT6VHVV-aQtHsmwYB7x:-MT7dnx34XWyp_ll5gcj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdd04f8e5bcfc50b46794910fea6dbcc1fa841c1/pkg/migration/migration.go#L17)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

> ```
> activate 
> ```
made active?
</blockquote></details>

Done.
___
*[pkg/migration/migration.go, line 21 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT6VPGH4qHjKNK3sEP2:-MT7dome-j-qnkNzOd7m:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdd04f8e5bcfc50b46794910fea6dbcc1fa841c1/pkg/migration/migration.go#L21)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

> ```
> pkg/sqlmigrations.
> ```
don't you want to get rid of it eventually? Perhaps drop mention of this?
</blockquote></details>

Done.
___
*[pkg/migration/migrationcluster/cluster.go, line 74 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT6mQ9q9klOtObUbetF:-MT6p9GfAf_G16FTOtt0:bcimfn7) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdd04f8e5bcfc50b46794910fea6dbcc1fa841c1/pkg/migration/migrationcluster/cluster.go#L74)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

i'm curious why do we want to rate limit here?  
</blockquote></details>

I didn't write this code, @irfansharif did. I'm not inclined to change it ([Jordan Lewis's law](https://twitter.com/JordanALewis/status/1246134064110866434))
___
*[pkg/migration/migrationjob/migration_job.go, line 78 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT6ZTBO8R54JyrG8xoG:-MT6pdUc8dJ4xRdDK7yB:b6vs2j2) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdd04f8e5bcfc50b46794910fea6dbcc1fa841c1/pkg/migration/migrationjob/migration_job.go#L78)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

So, that would mark the job completed?
</blockquote></details>

Yes but it won't mark the migration as completed. 
___
*[pkg/migration/migrationjob/migration_job.go, line 99 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT6XF-H62MboZlvQ-mS:-MT7ds2LDHD6IlnApk9v:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdd04f8e5bcfc50b46794910fea6dbcc1fa841c1/pkg/migration/migrationjob/migration_job.go#L99)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

nit: perhaps long hand?
I see you already discussed this... I don't know; most of the code does "if err != nil { return wrapf }"

</blockquote></details>

Done.
___
*[pkg/migration/migrationmanager/manager.go, line 269 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT6e6Zd2a4Nyo-_dkb0:-MT6plai96YHDKi483Jz:b-pi0tvu) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdd04f8e5bcfc50b46794910fea6dbcc1fa841c1/pkg/migration/migrationmanager/manager.go#L269)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

> ```
> migrationjob.CheckIfMigrationCompleted
> ```

nit: this almost feels like this method might have a better home inside migration package (or migrationmanager) instead of migrationjob?
</blockquote></details>

Heh it started in the `migrationmanager` package. Ultimately the job needs to write the completion. I suppose that the most correct thing to do would be to abstract it completely but that feels like overkill. 
___
*[pkg/migration/migrationmanager/manager.go, line 310 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT6iV8B9wvtrzjjc66F:-MT6pwn09EtVMLLTsFV4:b8rm1pi) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdd04f8e5bcfc50b46794910fea6dbcc1fa841c1/pkg/migration/migrationmanager/manager.go#L310)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

do you need to use cte here? Can't we just 
select id, status, crdb_internal.pb_to_json() pl
where status in ... AND pl->....?
</blockquote></details>

This isn't a CTE. This is a subquery. I'm not sure what you mean. In sql you can't refer to a projection column in a where clause.

```sql
demo@127.0.0.1:44909/movr> SELECT 1 as v WHERE v > 0;
ERROR: column "v" does not exist
SQLSTATE: 42703
demo@127.0.0.1:44909/movr> SELECT v FROM (SELECT 1 as v) WHERE v > 0;
  v
-----
  1
(1 row)

```
CTE would be:
```sql
WITH t AS (SELECT 1 AS v) SELECT v FROM t WHERE v > 0;
```
___
*[pkg/migration/migrationmanager/manager.go, line 341 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT6i49U3rI3di7kRObd:-MT6qbYg1oDHVy76Be9o:b-v4fn75) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdd04f8e5bcfc50b46794910fea6dbcc1fa841c1/pkg/migration/migrationmanager/manager.go#L341)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

will the cluster ever recover from this? 
</blockquote></details>

I believe it ends up just failing the job and sending a sentry report. This is an invariant, it should not happen. 
___
*[pkg/migration/migrationmanager/manager.go, line 360 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT6jDU76KEqcBhu907A:-MT6qjkd6c5CPghNhw76:bmy5eb2) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdd04f8e5bcfc50b46794910fea6dbcc1fa841c1/pkg/migration/migrationmanager/manager.go#L360)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

just wondering: can this be made more explicit? Like, perhaps tests (or test cluster?) can set a variable here that this manager is running under test and only then allow this? Otherwise panic or something?

Might be more trouble than worth though. 
</blockquote></details>

Done. Injected the testing knob inside the logic test that was depending on this behavior. I was sad about it too. 


<!-- Sent from Reviewable.io -->

irfansharif commented on pkg/sql/planner.go:
```diff
@@ -473,6 +474,11 @@ func (p *planner) DistSQLPlanner() *DistSQLPlanner {
 	return p.extendedEvalCtx.DistSQLPlanner
 }
 
+// MigrationRunDependencies returns the migration.Cluster if there is one.
```
Did we mean to say migration.Cluster here?

ajwerner replied:
Fixed. I was toying for a few minutes with what migrations will look like for tenants but backed off a bit.

irfansharif commented on pkg/migration/migration.go:
```diff
@@ -0,0 +1,259 @@
+}
+
+// RunDependencies are used by the job to run migrations.
+type RunDependencies interface {
```
Does "RunDependencies" match with terminology we use elsewhere in the jobs infrastructure? Or is the "Run" here referring the runnable migrations (KVMigrations.Run, SQLMigrations.Run)? I don't have a better name off-hand, though compared to the rest of the interfaces we've defined here, this was a bit less clear.

ajwerner replied:
I agree this was a wart. I've reworked it to be `migration.JobDeps` and pulled the `Registry` to just be defined inline. I think it's cleaner.

irfansharif commented on pkg/migration/migrationcluster/migrationcluster.go:
```diff
@@ -0,0 +1,85 @@
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package migrationcluster
```
[nit] Can we use a better file name for this file? At the moment we have this, and cluster.go, the latter of which houses the primary types in this sub package. 

Aside from ClusterConfig, everything here is liveness related? Should ClusterConfig move closer to migrationcluster.New instead?

ajwerner replied:
Removed this file. Moved the dependencies to `cluster.go` and the `NodesFromNodeLiveness` to `nodes.go`

irfansharif commented on pkg/sql/catalog/systemschema/system.go:
```diff
@@ -345,6 +345,17 @@ CREATE TABLE system.sqlliveness (
 )`
+
+	LongRunningMigrationsTableSchema = `
+CREATE TABLE system.long_running_migrations (
```
Probably not worth doing now, but if we're imagining that this subsystem will eventually grow to envelop pkg/sqlmigrations, maybe a better table name is in order? `internal_migrations`? 

Again, probably not worth introducing new names for, but curious what you think. I think it would also tie in better with the fact that we've named the top-level package as simply "migrations". I don't think we've used "long-running" anywhere in code before this PR.

ajwerner replied:
Done, I've removed "long running" in all of its forms. Everything is now just `migrations`.

irfansharif reviewed (approved):
Thanks for doing this! I've looked towards the parts of pkg/migrations I'm most comfortable with, and deferring to you + other reviewers for the bits I'm not (basically the jobs stuff).

irfansharif reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59760)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @miretskiy, and @tbg)

___
*[pkg/migration/migrationcluster/cluster.go, line 74 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59760#-MT6mQ9q9klOtObUbetF:-MT7lGsF9_xvRXKjJEng:b-mmjmeu) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdd04f8e5bcfc50b46794910fea6dbcc1fa841c1/pkg/migration/migrationcluster/cluster.go#L74)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I didn't write this code, @irfansharif did. I'm not inclined to change it ([Jordan Lewis's law](https://twitter.com/JordanALewis/status/1246134064110866434))
</blockquote></details>

The reasoning was to try and "roll-out" the operations run across every-node in the cluster, out of an abundance of caution. With a limit as high as 25(nodes in parallel), perhaps it's too high in practice. Can I cite Jordan's law for constants I introduced myself to avoid doing any work to change them?


<!-- Sent from Reviewable.io -->

ajwerner commented:
Created https://github.com/cockroachdb/cockroach/issues/60307 for a fast follow-on. Also fixed an issue with pausing and added a test.

ajwerner commented:
TFTR!

bors r+
